plugins/polio/api/rounds/reasons_for_delay.py

```diff
@@ -0,0 +1,95 @@
+from hat.audit.audit_mixin import AuditMixin
+from plugins.polio.models import ReasonForDelay
+from rest_framework import serializers
+from iaso.api.common import ModelViewSet, HasPermission
+from rest_framework import filters
+from rest_framework.decorators import action
+from rest_framework.response import Response
+from hat.menupermissions import models as permission
+from django_filters.rest_framework import DjangoFilterBackend  # type: ignore
+
+
+class AuditReasonForDelaySerializer(serializers.ModelSerializer):
+    class Meta:
+        model = ReasonForDelay
+        fields = "__all__"
+
+
+class ReasonForDelayForCampaignSerializer(serializers.ModelSerializer):
+    class Meta:
+        model = ReasonForDelay
+        fields = ["id", "name_fr", "name_en", "key_name"]
+
+
+class ReasonForDelayFieldSerializer(serializers.Field):
+    def to_representation(self, value):
+        if value:
+            return value.key_name
+        return None
+
+    def to_internal_value(self, data):
+        account = self.context["request"].user.iaso_profile.account
+        try:
+            reason_for_delay = ReasonForDelay.objects.get(key_name=data, account=account)
+            return reason_for_delay
+        except ReasonForDelay.DoesNotExist:
+            raise serializers.ValidationError(f"key_name not found: {data}")
+
+
+class ReasonForDelaySerializer(serializers.ModelSerializer):
+    class Meta:
+        model = ReasonForDelay
+        fields = ["id", "name_fr", "name_en", "key_name", "created_at", "updated_at", "times_selected"]
+
+        read_only_fields = ["created_at", "updated_at", "id"]
+
+    times_selected = serializers.SerializerMethodField()
+
+    def get_times_selected(self, reason_for_delay):
+        return len(list(reason_for_delay.round_history_entries.all()))
+
+    def validate(self, data):
+        key_name = data.get("key_name", None)
+        name_en = data.get("name_en", None)
+        if key_name is None:
+            raise serializers.ValidationError("requiredField")
+        if name_en is None:
+            raise serializers.ValidationError("requiredField")
+        # Checking that updates are made from user from the same accont as the Reason for delay
+        if self.instance is not None:
+            user_account = self.context["request"].user.iaso_profile.account
+            if user_account != self.instance.account:
+                raise serializers.ValidationError("wrongUserAccount")
```

Comment:
Checking perms/authorizations shouldn't be the responsability of the Serializer. Its purpose is to convert Python into JSON and vice versa.

This is better done in the ViewSet.

Reply:
fair enough

Reply:
It's also useless. I noticed, while writing tests, that trying to PATCH from the wrong account will result in a 404 before even reaching that code